dags/sentinel-2_data_transfer.py

```diff
@@ -71,100 +94,159 @@ def africa_tile_ids():
+
+    src_keys = []
+    for src_url in urls:
+        src_key = extract_src_key(src_url)
+        key_exists = s3_hook_oregon.check_for_key(
+            key, bucket_name=default_args["src_bucket_name"]
+        )
+        src_keys.append(src_key)
+
+        if not key_exists:
+            raise ValueError(
+                f"{key} does not exist in the {default_args['src_bucket_name']} bucket"
+            )
+
+    with ThreadPoolExecutor(max_workers=5) as executor:
+        task = {executor.submit(write_scene, key): key for key in src_keys}
+        for future in as_completed(task):
+            scene_to_copy = task[future]
+            try:
+                result = future.result()
+            except Exception as exc:
+                raise ValueError(f"{scene_to_copy} failed to copy")
+            else:
+                print(f"Task {scene_to_copy} succeeded with {result}")
+
 
-    end_index = min(start_index + num_msg_per_worker, len(messages)) \
-                    if index < last_worker_index \
-                    else len(messages)
+def is_valid_tile_id(message):
 
-    messages = messages[start_index : end_index]
-    attributes = ti.xcom_pull(key='attributes', task_ids='check_for_messages')
-    attributes = attributes[start_index : end_index]
+    body = json.loads(message.body)
+    metadata = json.loads(body["Message"])
+    tile_id = metadata["id"].split("_")[1]
 
-    # Load Africa tile ids
     valid_tile_ids = africa_tile_ids()
-    max_num_cpus = 10
-    pool = multiprocessing.Pool(processes=max_num_cpus, maxtasksperchild=2)
-    args = [(msg, atr, tile) for msg, atr, tile in zip(messages, attributes, [valid_tile_ids]*len(messages))]
-    results = pool.map(copy_scene, args)
-    Not_none_values = list(filter(None.__ne__, results))
+    if tile_id not in valid_tile_ids:
+        print(f"{tile_id} is not in the list of Africa tiles for {metadata.get('id')}")
+        return False
+    return True
 
-   # Update context with number of processed messages
-    ti.xcom_push(key='processed_msg_count', value=len(messages))
-    print(f"Copied {len(Not_none_values)} out of {len(messages)} files")
 
-def get_queue():
+def copy_s3_objects(ti, **kwargs):
     """
-    Return the SQS queue object
+    Copy objects from a s3 bucket to another s3 bucket.
+    :param ti: Task instance
     """
-    sqs_hook = SQSHook(aws_conn_id=dag.default_args['us_conn_id'])
-    queue_url = default_args['sqs_queue']
-    queue_name = queue_url[queue_url.rindex("/") + 1:]
-    sqs = sqs_hook.get_resource_type('sqs')
-    return sqs.get_queue_by_name(QueueName=queue_name)
+    successful = 0
+    failed = 0
+
+    sqs_hook = SQSHook(aws_conn_id=dag.default_args["us_conn_id"])
+    sqs = sqs_hook.get_resource_type("sqs")
+    queue = sqs.get_queue_by_name(QueueName=default_args.get("sqs_queue"))
+    messages = get_messages(queue, visibility_timeout=600)
+
+    for message in messages:
+        try:
+            if not is_valid_tile_id(message):
```

Comment:
And here we should pass in the list of tiles ids.

Reply:
Done.